hal/spi_api.h

```diff
@@ -19,220 +18,211 @@
+} spi_init_t;
 
-/**@}*/
 /**
- * \defgroup SynchSPI Synchronous SPI Hardware Abstraction Layer
- * @{
+ * This enumerates the possible result of the spi_init function.
  */
+typedef enum spi_result_t {
+    SPI_RESULT_OK, /**< Operation successful. */
+    SPI_RESULT_CONFIG_UNSUPPORTED, /**< The required parameters are not supported by the device. */
+    SPI_RESULT_INVALID_PARAM, /**< The given parameter(s) is/are invalid. */
+    SPI_RESULT_ALREADY_INITIALIZED, /**< The requested peripheral/SS pin is already initialized. */
+} spi_result_t;
 
-/** Write a byte out in master mode and receive a value
+/**
+ * This structure groups all required data to handle a SPI transaction.
  *
- * @param[in] obj   The SPI peripheral to use for sending
- * @param[in] value The value to send
- * @return Returns the value received during send
- */
-int  spi_master_write(spi_t *obj, int value);
-
-/** Write a block out in master mode and receive a value
- *
- *  The total number of bytes sent and received will be the maximum of
- *  tx_length and rx_length. The bytes written will be padded with the
- *  value 0xff.
- *
- * @param[in] obj        The SPI peripheral to use for sending
- * @param[in] tx_buffer  Pointer to the byte-array of data to write to the device
- * @param[in] tx_length  Number of bytes to write, may be zero
- * @param[in] rx_buffer  Pointer to the byte-array of data to read from the device
- * @param[in] rx_length  Number of bytes to read, may be zero
- * @param[in] write_fill Default data transmitted while performing a read
- * @returns
- *      The number of bytes written and read from the device. This is
- *      maximum of tx_length and rx_length.
- */
-int spi_master_block_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, char write_fill);
-
-/** Check if a value is available to read
+ * - If tx is NULL a fill symbol is sent instead. See spi_init_t.
+ * - If tx is not NULL, the tx buffer must be at least `tx_count` symbols long.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if a value is available
- */
-int  spi_slave_receive(spi_t *obj);
-
-/** Get a received value out of the SPI receive buffer in slave mode
+ * - If rx is NULL, all read symbols are dropped.
+ * - If rx is not NULL, the rx buffer must be at least `rx_count` symbols long.
  *
- * Blocks until a value is available
- * @param[in] obj The SPI peripheral to read
- * @return The value received
+ * A symbol might be bigger than a byte. In such case symbols are read/written following the
+ * platform's endianness.
  */
-int  spi_slave_read(spi_t *obj);
+typedef struct spi_transfer_args_s {
+    const uint8_t *tx; /**< A buffer containing the data to be sent. */
+    uint32_t tx_count; /**< The number of symbol to send. */
+    uint8_t *rx; /**< A buffer to store the received data. */
+    uint32_t rx_count; /**< The number of symbol to read. */
+} spi_transfer_args_t;
```

Comment:
In comments above, you have "arg" as singular in `spi_transfer_arg_t`, but here it's plural.

Reply:
Thank you !